### PR TITLE
TaskFlow-based tenant and image migration

### PR DESCRIPTION
**Pull request by antonf**
TaskFlow library make possible to create parallel restartable task flow system. After this patch it will be possible to migrate several images in parallel.

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3272/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.os.migrate.glance
cloudferry/lib/os/migrate/glance.py:32: [W0223(abstract-method), 
BaseImageMigrationTask] Method 'migrate' is abstract in class 'ObjectTask' but is not 
overridden
cloudferry/lib/os/migrate/glance.py:181: [W0631(undefined-loop-variable), 
UploadDeletedImage.upload_server_image] Using possibly undefined loop variable 'disk'
************* Module cloudferry.lib.os.discovery.model
cloudferry/lib/os/discovery/model.py:372: [E1101(no-member), Model.__eq__] Super of 
'Model' has no '__eq__' member
cloudferry/lib/os/discovery/model.py:779: [E1101(no-member), LazyObj.__eq__] Super of 
'LazyObj' has no '__eq__' member
************* Module cloudferry.lib.utils.local_db
cloudferry/lib/utils/local_db.py:49: [E1101(no-member), Transaction.__enter__] Instance 
of 'ThreadLocalStorage' has no 'depth' member
cloudferry/lib/utils/local_db.py:50: [E1101(no-member), Transaction.__enter__] Instance 
of 'ThreadLocalStorage' has no 'depth' member
cloudferry/lib/utils/local_db.py:70: [E1101(no-member), Transaction.__exit__] Instance 
of 'ThreadLocalStorage' has no 'depth' member

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.os.migrate.glance
cloudferry/lib/os/migrate/glance.py:32: [W0223(abstract-method), 
BaseImageMigrationTask] Method 'migrate' is abstract in class 'ObjectTask' but is not 
overridden
cloudferry/lib/os/migrate/glance.py:181: [W0631(undefined-loop-variable), 
UploadDeletedImage.upload_server_image] Using possibly undefined loop variable 'disk'
************* Module cloudferry.lib.os.discovery.model
cloudferry/lib/os/discovery/model.py:372: [E1101(no-member), Model.__eq__] Super of 
'Model' has no '__eq__' member
cloudferry/lib/os/discovery/model.py:779: [E1101(no-member), LazyObj.__eq__] Super of 
'LazyObj' has no '__eq__' member
************* Module cloudferry.lib.utils.local_db
cloudferry/lib/utils/local_db.py:49: [E1101(no-member), Transaction.__enter__] Instance 
of 'ThreadLocalStorage' has no 'depth' member
cloudferry/lib/utils/local_db.py:50: [E1101(no-member), Transaction.__enter__] Instance 
of 'ThreadLocalStorage' has no 'depth' member
cloudferry/lib/utils/local_db.py:70: [E1101(no-member), Transaction.__exit__] Instance 
of 'ThreadLocalStorage' has no 'depth' member

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferry/lib/os/discovery/nova.py [hunk 1]
```diff
@@ -153,13 +155,23 @@ def equals(self, other):
...
-    output = remote.sudo('virsh domblklist {instance}',
-                         instance=server.instance_name)
-    if not output.succeeded:
+    try:
+        output = remote_executor.sudo('virsh domblklist {instance}',
+                                      instance=server.instance_name)
+    except remote.RemoteFailure:
         LOG.warning('Unable to get ephemeral disks for server '
-                    '%s, skipping.', server.object_id)
+                    '%s, skipping.', server.object_id, exc_info=True)
```
Comment (SVilgelm) [inline cloudferry/lib/os/discovery/nova.py]: I suggest you to write a debug message with `exc_info` and remove it from warning.

File: cloudferry/lib/os/discovery/nova.py [hunk 2]
```diff
@@ -174,10 +186,12 @@ def _list_ephemeral(remote, server):
...
             continue
-        size_str = remote.sudo('stat -c %s {path}', path=path)
-        if not size_str.succeeded:
+        try:
+            size_str = remote_executor.sudo('stat -c %s {path}', path=path)
+        except remote.RemoteFailure:
             LOG.warning('Unable to get size of ephemeral disk "%s" for server '
-                        '%s, skipping disk.', path, server.object_id)
+                        '%s, skipping disk.', path, server.object_id,
+                        exc_info=True)
```
Comment (SVilgelm) [inline cloudferry/lib/os/discovery/nova.py]: ditto

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3274/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.os.migrate.glance
cloudferry/lib/os/migrate/glance.py:32: [W0223(abstract-method), 
BaseImageMigrationTask] Method 'migrate' is abstract in class 'ObjectTask' but is not 
overridden
cloudferry/lib/os/migrate/glance.py:181: [W0631(undefined-loop-variable), 
UploadDeletedImage.upload_server_image] Using possibly undefined loop variable 'disk'
************* Module cloudferry.lib.utils.local_db
cloudferry/lib/utils/local_db.py:49: [E1101(no-member), Transaction.__enter__] Instance 
of 'ThreadLocalStorage' has no 'depth' member
cloudferry/lib/utils/local_db.py:50: [E1101(no-member), Transaction.__enter__] Instance 
of 'ThreadLocalStorage' has no 'depth' member
cloudferry/lib/utils/local_db.py:70: [E1101(no-member), Transaction.__exit__] Instance 
of 'ThreadLocalStorage' has no 'depth' member

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.os.migrate.glance
cloudferry/lib/os/migrate/glance.py:32: [W0223(abstract-method), 
BaseImageMigrationTask] Method 'migrate' is abstract in class 'ObjectTask' but is not 
overridden
cloudferry/lib/os/migrate/glance.py:181: [W0631(undefined-loop-variable), 
UploadDeletedImage.upload_server_image] Using possibly undefined loop variable 'disk'
************* Module cloudferry.lib.utils.local_db
cloudferry/lib/utils/local_db.py:49: [E1101(no-member), Transaction.__enter__] Instance 
of 'ThreadLocalStorage' has no 'depth' member
cloudferry/lib/utils/local_db.py:50: [E1101(no-member), Transaction.__enter__] Instance 
of 'ThreadLocalStorage' has no 'depth' member
cloudferry/lib/utils/local_db.py:70: [E1101(no-member), Transaction.__exit__] Instance 
of 'ThreadLocalStorage' has no 'depth' member

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3275/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.utils.local_db
cloudferry/lib/utils/local_db.py:49: [E1101(no-member), Transaction.__enter__] Instance 
of 'ThreadLocalStorage' has no 'depth' member
cloudferry/lib/utils/local_db.py:50: [E1101(no-member), Transaction.__enter__] Instance 
of 'ThreadLocalStorage' has no 'depth' member
cloudferry/lib/utils/local_db.py:70: [E1101(no-member), Transaction.__exit__] Instance 
of 'ThreadLocalStorage' has no 'depth' member
************* Module cloudferry.lib.os.migrate.glance
cloudferry/lib/os/migrate/glance.py:34: [W0223(abstract-method), 
BaseImageMigrationTask] Method 'migrate' is abstract in class 'ObjectTask' but is not 
overridden
cloudferry/lib/os/migrate/glance.py:183: [W0631(undefined-loop-variable), 
UploadDeletedImage.upload_server_image] Using possibly undefined loop variable 'disk'

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.utils.local_db
cloudferry/lib/utils/local_db.py:49: [E1101(no-member), Transaction.__enter__] Instance 
of 'ThreadLocalStorage' has no 'depth' member
cloudferry/lib/utils/local_db.py:50: [E1101(no-member), Transaction.__enter__] Instance 
of 'ThreadLocalStorage' has no 'depth' member
cloudferry/lib/utils/local_db.py:70: [E1101(no-member), Transaction.__exit__] Instance 
of 'ThreadLocalStorage' has no 'depth' member

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferry/lib/os/cloud_db.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
+# Copyright (c) 2014 Mirantis Inc.
```
Comment (SVilgelm) [inline cloudferry/lib/os/cloud_db.py]: 2016?

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferry/lib/utils/remote.py [hunk 4]
```diff
@@ -104,11 +100,92 @@ def tmpdir(self, prefix='cloudferry'):
...
+            return None
+        client = paramiko.SSHClient()
+        client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        client.connect(
+            hostname=gateway.hostname,
+            port=gateway.port,
+            username=gateway.username,
+            password=gateway.password,
+            pkey=self._create_pkey(gateway.private_key),
+            sock=self._connect_through_gateway(
```
Comment (SVilgelm) [inline cloudferry/lib/utils/remote.py]: Looks like infinity loop? You just copied a `connect` function.

File: cloudferry/lib/utils/remote.py [hunk 5]
```diff
@@ -104,11 +100,92 @@ def tmpdir(self, prefix='cloudferry'):
...
+                saved_exceptions.append(sys.exc_info())
+
+        for exc_info in saved_exceptions:
+            LOG.error('Failed to decode SSH key', exc_info=exc_info)
+        return None
+
+    def _run(self, command, is_sudo):
+        LOG.debug('[%s] running command "%s"', self.hostname, command)
+        if is_sudo:
+            sudo_prompt = '[{magic}] sudo password:'.format(
```
Comment (SVilgelm) [inline cloudferry/lib/utils/remote.py]: nice 👍

File: cloudferry/lib/utils/remote.py (hunk 4 above)
Comment (antonf) [inline cloudferry/lib/utils/remote.py]: It done this way in order to connect to controller through several gateways (e.g. first through jump host, then through fuel master).

**Comment by SVilgelm** [comment]
👍 The functional test were stopped because we have no functional test fro this code. so, it can be merged.

**Pull request merged**